### ISSUE TITLE
Test Suite Cannot Resolve Multiply-Required Transitive Dependency

### ISSUE THREAD

**Issue by Matthew-Ielusic** [polyfy/polylith#108]
**Describe the bug**

`poly test` cannot run if two components have transitive dependecies on the same Github-hosted library.

**To Reproduce**
1. Clone a repo [I made to exhibit this behavior.](https://github.com/Matthew-Ielusic/MinimalBugRepro)
2. Run `poly test`
3. See error:
```Couldn't resolve libraries for the deploy project: clojure.lang.ExceptionInfo: Unable to compare versions for repro-lib/root: {:local/root "/Users/<etc.>```

**Expected behavior**
The test suite should run without issue - in fact, it does run without issue in Kaocha, through the script `kaocha.sh`.

**Operating System (please complete the following information):**
 - OS: MacOS
 - Version: 11.4 (20F71)

**Versions (please complete the following information):**
 - Java: 
```
openjdk 11.0.11 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)
```

 - Poly: 0.2.0-alpha11
 - Brew: 
```
Homebrew 3.2.6
Homebrew/homebrew-core (git revision fb7013ba7a; last commit 2021-08-11)
Homebrew/homebrew-cask (git revision d863760d7b; last commit 2021-08-10)
```

**Additional context**

To be clear:

- The repo I linked to has two components, `left` and `right`
- Those components depend on two different libraries [hosted in another repo I made](https://github.com/Matthew-Ielusic/ReproLibrary)
- Those two libraries each depend on a third library, hosted in the same repo
- I added that shared library to `:exclusions` in all the deps.edn files.

**Comment by tengstrand** [comment]
I'm on holiday right now, but will have a look at this in the weekend.

**Comment by seancorfield** [comment]
That exception is coming from `tools.deps.alpha` and it does not allow two git deps for the same dependency unless there is a direct ancestor link between them and the "newer" one can be directly inferred.

If you are using the CLI and `deps.edn` at all with transitive git deps, you are going to run into this -- it is not specific to Polylith -- and the CLI will fail to run your code in such a situation. The CLI (and `tools.deps.alpha`) always picks the newer dependency when computing the classpath. I'm not sure what Koacha is doing in your situation, but it is not following the behavior of the CLI and `deps.edn`.

**Comment by Matthew-Ielusic** [comment]
> If you are using the CLI and `deps.edn` at all with transitive git deps, you are going to run into this -- it is not specific to Polylith -- and the CLI will fail to run your code in such a situation. The CLI (and `tools.deps.alpha`) always picks the newer dependency when computing the classpath. I'm not sure what Koacha is doing in your situation, but it is not following the behavior of the CLI and `deps.edn`.

Is there a way to specify exclusions for git deps, then?

And the Kaocha script is just running `clojure -Srepro -M:test:dev -m kaocha.runner "$@"` - is Kaocha bypassing CLI deps resolving somehow?

**Comment by seancorfield** [comment]
> Is there a way to specify exclusions for git deps, then?

`:exclusions` are based on the lib name (group/artifact) -- whether they're on Maven, GitHub, local, etc is not relevant.

I've no idea what Kaocha is doing and I'd actually have to see your (partially) migrated codebase and compare it to your prior setup to see why/whether that mattered.

**Comment by imrekoszo** [comment]
I forked Matthew's projects to make the repro case even leaner:
https://github.com/imrekoszo/MinimalBugRepro/tree/no-common-transitive-deps
https://github.com/imrekoszo/ReproLibrary/tree/no-common-transitive-deps

In this setup, the `red` and `blue` libs in `ReproLibrary` do not share any dependencies with each other, yet the issue is still there with poly test.

`ReproLibrary` now contains 3 independent modules: `blue`, `green` (renamed from `root` as it's no longer a dependency of anything), `red` - all of these have 0 dependencies.

MinimalBugRepro contains 2 components: `left` (which test-time depends on `blue` alone) and `right` (which test-time depends on `red` alone). There is nothing in `MinimalBugRepro` that depends on `green` neither directly nor transitively.

The clojure tool runs fine next to each `deps.edn` in the project tree:

`left`'s dep tree:
```
> clojure -Srepro -Sforce -Stree -A:test
org.clojure/clojure 1.10.3
  . org.clojure/spec.alpha 0.2.194
  . org.clojure/core.specs.alpha 0.2.56
repro-lib/blue https://github.com/imrekoszo/ReproLibrary/ 4237c64
```

`right`'s dep tree:
```
> clojure -Srepro -Sforce -Stree -A:test
org.clojure/clojure 1.10.3
  . org.clojure/spec.alpha 0.2.194
  . org.clojure/core.specs.alpha 0.2.56
repro-lib/red https://github.com/imrekoszo/ReproLibrary/ 4237c64
```

project `deploy`'s dep tree:
```
> clojure -Srepro -Sforce -Stree
org.clojure/clojure 1.10.3
  . org.clojure/spec.alpha 0.2.194
  . org.clojure/core.specs.alpha 0.2.56
poly/left /Users/ikoszo/opensource/poly-deps-bug/MinimalBugRepro/components/left
poly/right /Users/ikoszo/opensource/poly-deps-bug/MinimalBugRepro/components/right
```

top-level dep tree:
```
> clojure -Srepro -Sforce -Stree -A:test:dev
org.clojure/clojure 1.10.3
  . org.clojure/spec.alpha 0.2.194
  . org.clojure/core.specs.alpha 0.2.56
repro-lib/blue https://github.com/imrekoszo/ReproLibrary/ 4237c64
repro-lib/red https://github.com/imrekoszo/ReproLibrary/ 4237c64
```

Yet, `poly test` still fails, complaining about not being able to compare versions for `green`, which, as stated above, shouldn't even be a dep:
```
> poly test
Projects to run tests from: deploy

Couldn't resolve libraries for the deploy project: clojure.lang.ExceptionInfo: Unable to compare versions for repro-lib/green: {:local/root "/Users/ikoszo/.gitlibs/libs/repro-lib/red/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", :deps/manifest :deps, :deps/root "/Users/ikoszo/.gitlibs/libs/repro-lib/red/4237c6492172dc6a4102ccfc34df6316ccd1734b/green"} and {:local/root "/Users/ikoszo/.gitlibs/libs/repro-lib/blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", :deps/manifest :deps, :deps/root "/Users/ikoszo/.gitlibs/libs/repro-lib/blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/green"} {:lib repro-lib/green, :coord-x {:local/root "/Users/ikoszo/.gitlibs/libs/repro-lib/red/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", :deps/manifest :deps, :deps/root "/Users/ikoszo/.gitlibs/libs/repro-lib/red/4237c6492172dc6a4102ccfc34df6316ccd1734b/green"}, :coord-y {:local/root "/Users/ikoszo/.gitlibs/libs/repro-lib/blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", :deps/manifest :deps, :deps/root "/Users/ikoszo/.gitlibs/libs/repro-lib/blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/green"}}
Unable to compare versions for repro-lib/green: {:local/root "/Users/ikoszo/.gitlibs/libs/repro-lib/red/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", :deps/manifest :deps, :deps/root "/Users/ikoszo/.gitlibs/libs/repro-lib/red/4237c6492172dc6a4102ccfc34df6316ccd1734b/green"} and {:local/root "/Users/ikoszo/.gitlibs/libs/repro-lib/blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", :deps/manifest :deps, :deps/root "/Users/ikoszo/.gitlibs/libs/repro-lib/blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/green"}
```

**Comment by seancorfield** [comment]
Thanks for the repro case @imrekoszo -- I notice that the case of the file paths doesn't seem to agree here:

```bash
Blue		README.md	Red		Root		deps.edn
(! 1055)-> cat deps.edn 
{:deps {repro-lib/blue {:local/root "blue"}
        repro-lib/red  {:local/root "red"}
        repro-lib/root {:local/root "root"}}}
```

I wonder if that's causing problems for tools.deps.alpha?

At this point, I think it may be worth asking this as a question on ask.clojure.org to see if there's an underlying t.d.a issue caused by case-sensitivity.

And a question to the Polylith folks: are :exclusions being respected when building test deps?

**Comment by seancorfield** [comment]
This may also be a restriction in t.d.a. that you can't have two git deps into the same rep with different `:deps/root` values. That seems the most likely thing, looking at the error.

**Comment by seancorfield** [comment]
I posted a link to this issue in the #tools-deps channel on Slack and asked Alex if he had any immediate insight into it.

**Comment by seancorfield** [comment]
After discussing this with Alex on Slack, it looks like this is related to https://clojure.atlassian.net/browse/TDEPS-132

Because the two git deps (in the top-level project) have different `:deps/root`, by the time the `:local/root` dep in that git project is resolved, it's full path depends on that deps root (as well as the overall git repo/SHA path) so the local deps are not comparable.

You can see that in the error message (reformatted and edited for clarity):

```
Unable to compare versions for repro-lib/green: 
{:local/root "/Users/sean/.gitlibs/libs/repro-lib/red/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", ...
and 
{:local/root "/Users/sean/.gitlibs/libs/repro-lib/blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/green", ...
```

These are "just" local deps and they point to different file paths (because of the red/blue deps root) so they are not comparable.

I suspect if you renamed the lib names in those local/root deps, so that you had repro-lib-red/green and repro-lib-blue/green, the problem would go away because t.d.a would not try to compare them -- because they would have unique lib names.

**Comment by imrekoszo** [comment]
@seancorfield thank you for spending time on this!

I probably wasn't fully clear on what changes I made in the `no-common-transitive-deps` branches in my forks:
( https://github.com/imrekoszo/MinimalBugRepro/tree/no-common-transitive-deps
https://github.com/imrekoszo/ReproLibrary/tree/no-common-transitive-deps )

1. fixed folder name casing (hopefully answering https://github.com/polyfy/polylith/issues/108#issuecomment-898649531)
2. renamed `root` to `green`
3. got rid of `:local/root` dependencies between red/green/blue (hopefully answering https://github.com/polyfy/polylith/issues/108#issuecomment-898690082)

And when I run the clojure cli in any of the projects in the poly repo (MinimalBugRepro), it runs fine and is able to resolve dependencies.

Deps in the lib monorepo:
```
> find . -name deps.edn
./green/deps.edn
./blue/deps.edn
./deps.edn
./red/deps.edn

> find . -name deps.edn | xargs -n 1 cat
{:paths ["src"]}

{:paths ["src"]}

{:deps {repro-lib/blue {:local/root "blue"}
        repro-lib/red  {:local/root "red"}
        repro-lib/green {:local/root "green"}}}

{:paths ["src"]}
```

Deps in the polylith project:
```
> find . -name deps.edn
./deps.edn
./projects/deploy/deps.edn
./components/right/deps.edn
./components/left/deps.edn

> find . -name deps.edn | xargs -n 1 cat
{:aliases  {:dev {:extra-paths ["development/src"
                                "components/left/src"
                                "components/right/src"]
                  :extra-deps {repro-lib/blue {:git/url "https://github.com/imrekoszo/ReproLibrary/"
                                               :sha "4237c6492172dc6a4102ccfc34df6316ccd1734b"
                                               :deps/root "blue"}
                               repro-lib/red  {:git/url "https://github.com/imrekoszo/ReproLibrary/"
                                               :sha "4237c6492172dc6a4102ccfc34df6316ccd1734b"
                                               :deps/root "red"}}}
            :test {:extra-paths ["components/left/test"
                                 "components/right/test"]}
            :kaocha {:extra-deps {lambdaisland/kaocha {:mvn/version "1.0.829"}}}}}

{:deps {poly/left {:local/root "../../components/left"}
        poly/right {:local/root "../../components/right"}}}

{:paths ["src" "resources"]
 :deps {}
 :aliases {:test {:extra-paths ["test"]
                  :extra-deps {repro-lib/red {:git/url "https://github.com/imrekoszo/ReproLibrary/"
                                              :sha "4237c6492172dc6a4102ccfc34df6316ccd1734b"
                                              :deps/root "red"}}}}}

{:paths ["src" "resources"]
 :deps {}
 :aliases {:test {:extra-paths ["test"]
                  :extra-deps {repro-lib/blue {:git/url "https://github.com/imrekoszo/ReproLibrary/"
                                               :sha "4237c6492172dc6a4102ccfc34df6316ccd1734b"
                                               :deps/root "blue"}}}}}
```

**Comment by imrekoszo** [comment]
I did find something however. If you look at this part from my previous comment:

Deps in the lib monorepo:
```
> find . -name deps.edn
./green/deps.edn
./blue/deps.edn
./deps.edn
./red/deps.edn

> find . -name deps.edn | xargs -n 1 cat
{:paths ["src"]}

{:paths ["src"]}

{:deps {repro-lib/blue {:local/root "blue"}
        repro-lib/red  {:local/root "red"}
        repro-lib/green {:local/root "green"}}}

{:paths ["src"]}
```

The lib monorepo has a top-level `deps.edn`, referencing all 3 of the sub-projects. When I removed those references, poly test started working: https://github.com/imrekoszo/ReproLibrary/tree/no-top-deps and https://github.com/imrekoszo/MinimalBugRepro/tree/no-top-deps

After the change, deps in the lib monorepo look like:
```
> find . -name deps.edn
./green/deps.edn
./blue/deps.edn
./deps.edn
./red/deps.edn

> find . -name deps.edn | xargs -n 1 cat
{:paths ["src"]}

{:paths ["src"]}

{:deps {}}

{:paths ["src"]}
```


The strange thing is, that nothing in MinimalBugRepro depends on that top-level `deps.edn` in ReproLibrary. What I suspect is that the poly tool takes that into account by mistake. I could be wrong however.

**Comment by seancorfield** [comment]
I made sure to use the `no-common-transitive-deps` for the repro when discussing with Alex. The case issue was in the original repo, as you noted, and I had tried that first out of curiosity since I wanted to eliminate that first.

That top `deps.edn` is definitely part of the problem and you can see from how the lib repo ends up in the `.gitlibs` cache:

```bash
(! 1117)-> pwd
/Users/sean/.gitlibs/libs/repro-lib
(! 1118)-> ls
blue	red
(! 1119)-> cat blue/4237c6492172dc6a4102ccfc34df6316ccd1734b/deps.edn 
{:deps {repro-lib/blue {:local/root "blue"}
        repro-lib/red  {:local/root "red"}
        repro-lib/green {:local/root "green"}}}
(! 1120)-> cat red/4237c6492172dc6a4102ccfc34df6316ccd1734b/deps.edn 
{:deps {repro-lib/blue {:local/root "blue"}
        repro-lib/red  {:local/root "red"}
        repro-lib/green {:local/root "green"}}}
```

Those local dep libs are going to have different filesystem paths for the "same" lib name (`repro-lib/green`) which is why they're not comparable.

So, yes, it does look like something `poly test` is doing with `tools.deps.alpha` is incorrect because it should not be using that top `deps.edn` -- it should only use the `deps.edn` from the `:deps/root` inside that lib -- but I would not have expected `poly` to be following transitive deps at all _outside the workspace_.

However, the _original_ repo did have cross-dependencies in the subprojects and I believe, based on what Alex said, that it could still fall foul of the TDEP-132 bug mentioned above, unless the cross-dependencies are given unique lib names (as I suggested with `repro-lib-red/green` and `repro-lib-blue/green`).

In other words, I think there are two issues interacting here. I'll take a look at what `poly test` is doing and see if I can figure out what it is doing with t.d.a that causes the issue in the minimal repro with no common transitive deps.

**Comment by imrekoszo** [comment]
> However, the _original_ repo did have cross-dependencies in the subprojects and I believe, based on what Alex said, that it could still fall foul of the TDEP-132 bug mentioned above, unless the cross-dependencies are given unique lib names (as I suggested with `repro-lib-red/green` and `repro-lib-blue/green`).

Agreed, TDEPS-132 is certainly involved in the original problem. In fact, that behavior is something we need to deal with on a daily basis :). While annoying, it can fortunately be managed with direct deps and exclusions in our experience.

> In other words, I think there are two issues interacting here. I'll take a look at what `poly test` is doing and see if I can figure out what it is doing with t.d.a that causes the issue in the minimal repro with no common transitive deps.

Thank you very much for that!

**Comment by seancorfield** [comment]
My hypothesis is that `poly test` is not handling `:deps/root` correctly when building deps. Still debugging.

**Comment by seancorfield** [comment]
I have a fix for this locally -- I'll get a PR together shortly:

```bash
(! 1141)-> clojure -M:poly test
Projects to run tests from: deploy

Running tests from the deploy project, including 2 bricks: left, right

Testing matthew-ielusic.left.interface-test

Ran 1 tests containing 1 assertions.
0 failures, 0 errors.

Test results: 1 passes, 0 failures, 0 errors.

Testing matthew-ielusic.right.interface-test

Ran 1 tests containing 1 assertions.
0 failures, 0 errors.

Test results: 1 passes, 0 failures, 0 errors.
Execution time: 1 seconds
```
I have not yet tested this on the _original_ repo...

**Comment by seancorfield** [comment]
The original repo is still broken because `poly` doesn't respect `:exclusions` on git deps -- with that fix in place, the original repo can also be tested. I'll ensure my PR includes both fixes.

**Comment by imrekoszo** [comment]
Good man @seancorfield, much appreciated!

**Comment by seancorfield** [comment]
PR #111 is up. Adds support for `:exclusions` on all dep types (I checked with Alex that is correct, even though it is only documented for Maven) and adds support for `:deps/root` to git deps. Both of these are needed to fix @Matthew-Ielusic 's original problem (and his git deps need at least one exclusion on that common transitive dep, as far as I can tell for TDEPS-132).

I also updated the git dep handling to support `:git/tag` and `:git/sha` (as well as `:tag` and `:sha`) per the latest `tools.deps.alpha` (which Polylith uses), and made `:git/url` optional since t.d.a can deduce it from the lib-name if it is a com.github/io.github (or gitlab) style lib-name now.

**Comment by tengstrand** [comment]
Thanks @seancorfield for all this!
The PR is merged into `master`.

And to answer you question, yes, `:exclusions` are being respected when executing the `test` command.
You can run the `libs` command to verify that and to check which versions of the libraries it uses for each project. 

Another way is to execute e.g. `poly ws get:projects:dev:lib-deps` to check the libraries for the `dev` project
(both project name and alias can be used). It will return the same list as you see when executing the `libs` command,
because the internal `ws` structure is always calculated and then the different commands operate on that data structure.

**Comment by seancorfield** [comment]
@Matthew-Ielusic can you confirm that if you use the latest git version of the `poly` tool, your problem is now solved (with the appropriate `:exclusions` in place)?

**Comment by Matthew-Ielusic** [comment]
> @Matthew-Ielusic can you confirm that if you use the latest git version of the `poly` tool, your problem is now solved (with the appropriate `:exclusions` in place)?

Yes, it works!  Thank you.

**Comment by tengstrand** [comment]
Okay, cool!
